### DISCUSSION TITLE
GridPatch suggestion/help (NIfTI into patches)

### DISCUSSION THREAD

**Discussion by lenlobo** [Project-MONAI/MONAI#6334] [Q&A]
Hi,
I am trying to divide a large volume of a nifti into patches with some overlap, is there a tutorial or snippet that might help me with this? I was looking into the modules and closest definition resembles GridPatch and hence the title. My nifti is about 40gigs and to use this in monailabel without sampling/patches would be very resource hungry for me.

**Comment by KumoLiu** [comment]
Hi @lenlobo, you could find some examples in this [tutorial](https://github.com/Project-MONAI/tutorials/blob/main/pathology/multiple_instance_learning/panda_mil_train_evaluate_pytorch_gpu.py).
Also, for how to generate patches from the images you could also referring to these tutorials.
https://github.com/Project-MONAI/tutorials/blob/main/modules/2d_slices_from_3d_training.ipynb
https://github.com/Project-MONAI/tutorials/blob/main/modules/2d_slices_from_3d_sampling.ipynb

Hope it can help you, thanks!

**Reply by lenlobo** [reply]
Hi @KumoLiu ,

Thanks for the reply, I'll give them a look. I did in the mean time tried something and guess I figured it out. 
:D

**Reply by lenlobo** [reply]
Hey @KumoLiu ,

https://github.com/Project-MONAI/MONAI/discussions/3707#discussioncomment-4706698

I had tried this suggestion, and it was your comment itself. So with GridPatch, is there a way to save the information of the coordinates of the starting and ending of the slice? Lets say 500x500x500, is divided into 5 patches of 100 patch size, 500x100x500 with 0.2 overlap and so on. Is there a way to store the slice information retaining the coordinates from the original volume? 
I want to achieve this is because, I am 3d rendering the cropped volumes in Slicer and want to see them all together, but due to them having same pixel coordinates, all the renders are overlapping each other and I can't move them.

Thanks!

**Reply by KumoLiu** [reply]
Hi @lenlobo, I think `GridPatch` can preserve patch location if you are processing `MetaTensor`.
https://github.com/Project-MONAI/MONAI/blob/57c618cc0be2c081bae4caec13cf326c66874605/monai/transforms/spatial/array.py#L3017-L3024

Try to use `data.meta.keys()` to check if the information have been saved.
Hope it can help you, thanks!

**Reply by lenlobo** [reply]
So @KumoLiu ,
I got some information from data.meta, particularly I think the location key/tag helps me here, as I am mentioning patch size 100 in GridPatch, I have location as 0,100,200.. My scenario here is, If I upload all the cropped data or patches together to 3D Slicer, and 3d volume render them, they will overlap on top of each other, is there a way I can reconstruct the image together, all sliced volumes side by side in 3d? That is the reason I wanted the slice location but I am not sure how to use it.

**Reply by KumoLiu** [reply]
Hi @lenlobo, I would suggest you refer to the `aggregate` function in this class.
https://github.com/Project-MONAI/MONAI/blob/b356fecdb265ee9af4ce91a0c8238731994b5095/monai/inferers/merger.py#L101

Hope it can help you, thanks!

**Reply by lenlobo** [reply]
Hey @KumoLiu 
I tried two systems, my personal and work, and it seems AvgMerger, or Merger, both dont seem to be available? am I doing something wrong here?
![image](https://user-images.githubusercontent.com/46513163/232338795-9f39eaeb-42d2-4806-bb15-1f5a8db22598.png)

**Reply by KumoLiu** [reply]
Hi @lenlobo, yes, it's a new feature available in MONAI1.2, you could just referring it to [here](https://github.com/Project-MONAI/MONAI/blob/dev/monai/inferers/merger.py) or try a tag 1.2.0rc4.
Thanks!

**Reply by lenlobo** [reply]
Hi @KumoLiu , thanks for the help. I was able to get 3d patches of 100 slices each with GridPatch, the issue I am right now facing is, if I load two patches at the same time into lets say 3d slicer, they are overlapping completely on each other, my supervisor mentioned this is due to them having the same affines, as the original image, which indeed is the case. Is there a way to use GridPatch (or something else) and then create patches and also each patch would have a different affine based on from where it was cropped from in respect to the original volume?
My goal here would be to first, patch or crop the volume, and then observe them on Slicer and have them load together on that reference space side by side, not overlapping each other.

**Reply by KumoLiu** [reply]
Hi @lenlobo, if you are using something like `GridPatch` to generate several patches from a large image, they must be the same affine because they are from the same image and have the same resolution, right?
And IMO if you want to visualize something like the prediction results from these patches, you may need to merge these patches into a large image just like what it used to be? So I would suggest you to use the `aggregate` function in the `AvgMerger` which can help you merge patches by taking average of the overlapping area.
I'm not sure whether there is a way to load 100 slices simultaneously and display them in Slicer. Maybe you could try to use `matshow3d`. 
https://github.com/Project-MONAI/MONAI/blob/71578328732e3b7b95d90b2091c0976b8bc151ab/monai/visualize/utils.py#L36
There are some using examples you could find in this [tutorial](https://github.com/Project-MONAI/tutorials/blob/main/modules/transform_visualization.ipynb).

Hope it can help you!
Thanks!

**Reply by lenlobo** [reply]
@KumoLiu , So I am trying to use MONAI Label, and I have a remote server with some data. Some files are fine to load directly and start annotating, but some are too large and I need to crop them. Hence, I was using GridPatch to create these crops. In my current code, I use GridPatch and create lets say, patches of size 100, so I end up with around 7-8 tensors when the dimensions are [x,800,y].
I then use nibabel and save these tensors as nifti files and observe them in Slicer. 
Now before saving to nifti files, if I check the affines on a certain slice, it is the same as the original volume. Is there a way to ensure that when a slice is generated it has a different affine in relation to the original volume?  For me I am trying to reach a stage where I can load all the slices together and basically see a whole (looks reconstructed but not really connected) image on Slicer. Do you think it is possible?

**Reply by KumoLiu** [reply]
Hi @lenlobo, I'm not familiar with the Slicer but I think ITK can't load and view several patches in the same screen. I think maybe you want to check your data, then just try `matshow3d`.
Or if you want to modify the affine, you could do something manually like:
```
import numpy as np
from monai.transforms import LoadImage

data = LoadImage(image_only=True)(data_path)

print(data.affine) # --> original affine
data.meta["affine"] = np.eye(4)
print(data.affine) --> modified affine
```
Thanks!
